Bluetooth: BR: Add listener cb for discovery

lylezhu2012 opened the pull request:
The old [PR](https://github.com/zephyrproject-rtos/zephyr/pull/74577) was closed due to the source branch has been deleted. Create a new PR for the changes.

The results of inquiry and extended inquiry are only reported from application level after the inquiry complete event notified.

While the event of inquiry result is notified by controller in real time.

It is not a good user experience.

Just like scanning of LE, add a listener cb for discovery.

When the event of inquiry result, extended inquiry result, or remote name request complete notified, call listener recv cb to notify the upper layer.

When the event of inquiry complete notified or no pending of remote name request , call listener timeout cb to notify the upper layer.

lylezhu2012 committed:
Bluetooth: BR: Add listener cb for discovery

The results of inquiry and extended inquiry
are only reported from application level
after the inquiry complete event notified.

While the event of inquiry result is notified
by controller in real time.

It is not a good user experience.

Just like scanning of LE, add a listener
cb for discovery.

When the event of inquiry result, extended
inquiry result, or remote name request
complete notified, call listener `recv`
cb to notify the upper layer.

When the event of inquiry complete
notified or no pending of remote name
request , call listener `timeout`
cb to notify the upper layer.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: classic: Move classic source to sub folder

Move source code of classic from bluetooth.h to
./classic/classic.h.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: Remove callback from bt_br_discovery_start

Remove BR discovery callback from bt_br_discovery_start.

All discovery results will be notified through callback
registered by bt_br_discovery_cb_register.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
sample: Bluetooth: hfp_ag: update device discovery

Due to the update of function bt_br_discovery_start,
register discovery callback by calling
bt_br_discovery_cb_register.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: shell: BR: update device discovery

Due to the update of function bt_br_discovery_start,
register discovery callback by calling
bt_br_discovery_cb_register.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -0,0 +1,197 @@
+
+/** @brief BR/EDR discovery result structure */
+struct bt_br_discovery_result {
+	/** private */
+	uint8_t _priv[4];
```
I was just looking at how this is used, and it seems quite fragile. I'd recommend to clean it up in the same go. These four bytes are cast into the internally defined `struct discovery_priv`, but there's not even any build assert to ensure that the sizes match up. I'd suggest to just put the private struct in the public header, but cleanly document that it's private data, i.e. something like:
```c
struct bt_br_discovery_result {
	struct bt_br_discovery_priv _priv;
	...
};
```

lylezhu2012 replied:
Updated.

lylezhu2012 committed:
Bluetooth: BR: Optimize _priv of struct bt_br_discovery_result

Move `struct discovery_priv` to classic.h

Rename `struct discovery_priv` to `struct bt_br_discovery_priv`.

Modify the structure `struct bt_br_discovery_priv` with `@private`.

Change field `_priv` of `struct bt_br_discovery_result` from
`uint8_t _priv[4]` to `struct bt_br_discovery_priv _priv`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -38,10 +38,20 @@ extern "C" {
  * @{
  */
 
+/** @brief BR/EDR discovery private structure */
```
Either drop the doxygen annotation from the comments, or mark this as hidden or private (there's dedicated oxygen syntax for it, IIRC)

jhedberg commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -38,10 +38,20 @@ extern "C" {
+	uint8_t pscan_rep_mode;
+	/** Resolving remote name*/
+	uint8_t resolving;
+} __packed;
```
I don't think you need the `__packed`. It's also confusing, since this is normally reserved for protocol structs.

jhedberg commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -38,10 +38,20 @@ extern "C" {
+	/** Page scan repetition mode */
+	uint8_t pscan_rep_mode;
+	/** Resolving remote name*/
+	uint8_t resolving;
```
Should this be a `bool`?

lylezhu2012 committed:
Bluetooth: BR: Change the type of `resolving` to `bool`

The `resolving` is used to flag the status that the stack is requesting
remote device name.

It is better to use the type `bool` instead of the original type
`uint8_t`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>
